File: Libraries/LibWeb/Layout/ReplacedWithChildrenFormattingContext.cpp
/*
 * Copyright (c) 2026, Gregory Bertilson <gregory@ladybird.org>
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <LibWeb/Layout/BlockFormattingContext.h>
#include <LibWeb/Layout/ReplacedWithChildrenFormattingContext.h>

namespace Web::Layout {

ReplacedWithChildrenFormattingContext::ReplacedWithChildrenFormattingContext(LayoutState& state, LayoutMode layout_mode, Box const& box, FormattingContext* parent)
    : FormattingContext(Type::ReplacedWithChildren, layout_mode, state, box, parent)
{
}

void ReplacedWithChildrenFormattingContext::run(LayoutInput const& layout_input)
{
    auto const& available_space = layout_input.available_space;
    auto& root_state = m_state.get_mutable(context_box());
    auto content_inline_size = root_state.content_inline_size();

    // Mark the replaced element as having definite dimensions when the parent FC has
    // computed them from intrinsic size, so children with percentage sizes can resolve.
    auto natural_size = context_box().natural_size();
    if (natural_size.has_width())
        root_state.set_has_definite_inline_size(true);
    if (natural_size.has_height())
        root_state.set_has_definite_block_size(true);

    // For the block axis, use the parent-set content block size if it has been resolved (e.g. an
    // explicit or intrinsic block size); otherwise use the parent formatting context's space.
    auto child_available_block_size = root_state.has_definite_block_size()
        ? AvailableSize::make_definite(root_state.content_block_size())
        : available_space.block_size;

    auto child_available_space = AvailableSpace(
        AvailableSize::make_definite(content_inline_size),
        child_available_block_size);

    // The TreeBuilder wraps shadow DOM children in an anonymous BlockContainer.
    // Delegate layout to a BFC for that wrapper.
    auto const* wrapper = context_box().first_child_of_type<BlockContainer>();
    if (!wrapper)
        return;

    auto wrapper_constraints = constraints_for_child_context(root_state, layout_input.containing_block_constraints);
    auto& wrapper_state = m_state.create(*wrapper, wrapper_constraints.percentage_basis_inline_size, wrapper_constraints.percentage_basis_block_size);
    wrapper_state.set_content_inline_size(content_inline_size);

    auto bfc = make<BlockFormattingContext>(m_state, m_layout_mode, *wrapper, this);
    bfc->run(LayoutInput { child_available_space, wrapper_constraints });

    m_automatic_content_inline_size = content_inline_size;
    m_automatic_content_block_size = bfc->automatic_content_block_size();

    wrapper_state.set_content_block_size(m_automatic_content_block_size);

    place_child(*wrapper, { 0, 0 });

    bfc->parent_context_did_dimension_child_root_box();
}

CSSPixels ReplacedWithChildrenFormattingContext::automatic_content_inline_size() const
{
    return m_automatic_content_inline_size;
}

CSSPixels ReplacedWithChildrenFormattingContext::automatic_content_block_size() const
{
    return m_automatic_content_block_size;
}

void ReplacedWithChildrenFormattingContext::parent_context_did_dimension_child_root_box()
{
    layout_absolutely_positioned_children();
}

}

File: Libraries/LibWeb/Rust/src/layout/replaced_with_children_formatting_context.rs
/*
 * Copyright (c) 2026, Gregory Bertilson <gregory@ladybird.org>
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

use super::*;

pub(super) fn layout_replaced_with_children(
    run: &formatting_context::FormattingContextRun,
    layout_input: geometry::LayoutInput,
) -> formatting_context::ChildLayoutResult {
    // The parent FC has already resolved this replaced box's used size (natural size, explicit size, or the default
    // object size), so both dimensions are definite for its children — shadow content sized to fill
    // (e.g. width/height: 100%) resolves against them.
    let (content_inline_size, root_content_block_size) = {
        let root_state = run.records.used_values(run.box_);
        root_state.has_definite_inline_size.set(true);
        root_state.has_definite_block_size.set(true);
        (
            root_state.content_inline_size.get(),
            root_state.content_block_size.get(),
        )
    };

    let child_available_space = geometry::AvailableSpace {
        inline_size: geometry::AvailableSize::definite(content_inline_size),
        block_size: geometry::AvailableSize::definite(root_content_block_size),
    };

    // The TreeBuilder wraps shadow DOM children in an anonymous BlockContainer.
    // Delegate layout to a BFC for that wrapper.
    let mut wrapper = run.callbacks.first_child(run.box_);
    while !wrapper.is_invalid() {
        if node_facts::NodeFacts::new(&run.callbacks, wrapper).is_block_container() {
            break;
        }
        wrapper = run.callbacks.next_sibling(wrapper);
    }
    if wrapper.is_invalid() {
        return formatting_context::ChildLayoutResult::default();
    }

    let wrapper_constraints = run
        .sizing()
        .constraints_for_child_context(run.box_, layout_input.containing_block_constraints);
    let wrapper_state = run
        .records
        .create_used_values(&run.callbacks, wrapper, wrapper_constraints);
    wrapper_state.set_content_inline_size(content_inline_size);

    let wrapper_result = formatting_context::run_formatting_context(
        run.purpose,
        run.fragments.as_deref(),
        &wrapper_state,
        wrapper,
        None,
        formatting_context::FfiFormattingContextType::Block,
        run.layout_mode,
        run.should_collect_devtools_layout_data,
        run.callbacks,
        geometry::LayoutInput {
            available_space: child_available_space,
            containing_block_constraints: wrapper_constraints,
            content_box_position_in_bfc_root: None,
            sizing: geometry::RootSizingDirectives {
                adopt_automatic_content_block_size: true,
                ..geometry::RootSizingDirectives::default()
            },
            participation: geometry::ParticipationInParentFormattingContext::Item,
        },
        None,
    );
    formatting_context::propagate_percentage_block_size_dependency_to_containing_block(
        &run.records,
        &run.callbacks,
        wrapper,
        wrapper_result.depends_on_percentage_block_size,
    );
    formatting_context::place_child(run, wrapper, used_values::FfiCssPixelPoint::default(), None);

    formatting_context::ChildLayoutResult {
        automatic_content_inline_size: content_inline_size,
        automatic_content_block_size: wrapper_result.automatic_content_block_size,
        baselines: formatting_context::DerivedBaselines::default(),
        ..formatting_context::ChildLayoutResult::default()
    }
}
